airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py

```diff
@@ -164,54 +197,67 @@ def _read_pages(
+        }
+        while True:
+            chunk_id = self._next_chunk_id(property_chunks)
+            if chunk_id is None:
+                # pagination complete
+                break
 
-            records = {}
+            property_chunk = property_chunks[chunk_id]
+            request, response = self._fetch_next_page_for_chunk(
+                stream_slice, stream_state, property_chunk.next_page, property_chunk.properties
+            )
 
-            if not any(next_pages.values()):
-                pagination_complete = True
+            # When this is the first time we're getting a chunk's records, we set this to False to be used when deciding the next chunk
+            if property_chunk.first_time:
+                property_chunk.first_time = False
+            property_chunk.next_page = self.next_page_token(response)
+            chunk_page_records = records_generator_fn(request, response, stream_state, stream_slice)
+            if not self.too_many_properties:
+                # this is the case when a stream has no primary key
+                # (it is allowed when properties length does not exceed the maximum value)
+                # so there would be a single chunk, therefore we may and should yield records immediately
+                for record in chunk_page_records:
+                    property_chunk.record_counter += 1
+                    yield record
+                continue
+
+            # stick together different parts of records by their primary key and emit if a record is complete
+            for record in chunk_page_records:
+                property_chunk.record_counter += 1
+                record_id = record[self.primary_key]
+                if record_id not in records_by_primary_key:
+                    records_by_primary_key[record_id] = (record, 1)
+                    continue
+                partial_record, counter = records_by_primary_key[record_id]
+                partial_record.update(record)
+                counter += 1
+                if counter == len(property_chunks):
+                    yield partial_record  # now it's complete
+                    records_by_primary_key.pop(record_id)
+                else:
+                    records_by_primary_key[record_id] = (partial_record, counter)
+
+        # Process what's left.
+        # Because we make multiple calls to query N records (each call to fetch X properties of all the N records),
+        # there's a chance that the number of records corresponding to the query may change between the calls.
+        # Select 'a', 'b' from table order by pk -> returns records with ids `1`, `2`
+        #   <insert smth.>
+        # Select 'c', 'd' from table order by pk -> returns records with ids `1`, `3`
+        # Then records `2` and `3` would be incomplete.
+        # This may result in data inconsistency. We skip such records for now and log a warning message.
```

Comment:
So now that we're able to stitch records together, is this comment still relevant? Do you have a sense for what might cause us to still have partial records. I think I agree with keeping this in for visibility just in case, but this does have me curious

Reply:
Originally this was added to catch records that might be inserted between the two api calls. I dont think it's a common case, but we'd better have it to know if the number of such records becomes unreasonably high